gosrc2cpg: structure support

hiteshbedre opened the pull request:
Changes:
1. Added support for Go's structure.
2. Covered test cases having simple structure definition to complex(one structure used under another)

DavidBakerEffendi commented on joern-cli/frontends/gosrc2cpg/src/main/scala/io/joern/gosrc2cpg/astcreation/AstForExpressionCreator.scala:
```diff
@@ -23,24 +23,14 @@ trait AstForExpressionCreator(implicit withSchemaValidation: ValidationMode) { t
+        val (typeFullName, typeFullNameForcode, isVariadic, evaluationStrategy) = processTypeInfo(typeInfo)
+        val fieldNodeInfo = createParserNodeInfo(field.json(ParserKeys.Names).arr.head)
+        val fieldName     = fieldNodeInfo.json(ParserKeys.Name).str
+        Ast(memberNode(typeInfo, fieldName, s"${fieldName} ${typeFullNameForcode}", typeFullName, Seq()))
```
```suggestion
        Ast(memberNode(typeInfo, fieldName, s"$fieldName $typeFullNameForCode", typeFullName, Seq()))
```

hiteshbedre replied:
Addressed  here: https://github.com/joernio/joern/pull/3567/commits/7939c8bca109bad436a02b51d50e84e12403ce1d

DavidBakerEffendi commented on joern-cli/frontends/gosrc2cpg/src/main/scala/io/joern/gosrc2cpg/astcreation/AstForExpressionCreator.scala:
```diff
@@ -23,24 +23,14 @@ trait AstForExpressionCreator(implicit withSchemaValidation: ValidationMode) { t
       case Field => {
-        Ast()
+        val typeInfo = createParserNodeInfo(field.json(ParserKeys.Type))
+        val (typeFullName, typeFullNameForcode, isVariadic, evaluationStrategy) = processTypeInfo(typeInfo)
```
```suggestion
        val (typeFullName, typeFullNameForCode, isVariadic, evaluationStrategy) = processTypeInfo(typeInfo)
```

hiteshbedre replied:
Addressed here: https://github.com/joernio/joern/pull/3567/commits/7939c8bca109bad436a02b51d50e84e12403ce1d

DavidBakerEffendi commented on joern-cli/frontends/gosrc2cpg/src/main/scala/io/joern/gosrc2cpg/astcreation/AstForStructuresCreator.scala:
```diff
@@ -0,0 +1,70 @@
+      case Some(typeSpec) =>
+        val nameNode          = typeSpec.json(ParserKeys.Name)
+        val typeNode          = createParserNodeInfo(typeSpec.json(ParserKeys.Type))
+        val a: NewTypeDecl    = methodAstParentStack.collectFirst { case t: NewTypeDecl => t }.get
```
Could we make this
```scala
methodAstParentStack.collectFirst { case t: NewTypeDecl => t } match
  Some(a) => ...
  None => Seq.empty
```

hiteshbedre replied:
Addressed here: https://github.com/joernio/joern/pull/3567/commits/7939c8bca109bad436a02b51d50e84e12403ce1d

hiteshbedre commented:
@DavidBakerEffendi Addressed code review comment in: https://github.com/joernio/joern/pull/3567/commits/7939c8bca109bad436a02b51d50e84e12403ce1d

hiteshbedre commented:
New addressed changes:
1. Updated existing test cases in more structured way(using `in` clause)
2. Added test cases scenarios where we are moving member node to typeDecl node.

DavidBakerEffendi commented on joern-cli/frontends/gosrc2cpg/src/main/scala/io/joern/gosrc2cpg/astcreation/AstForStructuresCreator.scala:
```diff
@@ -0,0 +1,73 @@
+      case Some(typeSpec) =>
+        val nameNode = typeSpec.json(ParserKeys.Name)
+        val typeNode = createParserNodeInfo(typeSpec.json(ParserKeys.Type))
+        val a = methodAstParentStack.collectFirst { case t: NewTypeDecl => t } match {
```
I meant the whole block below should be in the `Some` part. e.g. 
```scala
methodAstParentStack.collectFirst { case t: NewTypeDecl => t } match {
   Some(a) => 
        val astParentType     = a.label
         val astParentFullName = a.fullName
         val fullName          = nameNode(ParserKeys.Name).str // TODO: Discuss fullName structure
         val typeDeclNode_ =
           typeDeclNode(
             typeSpec,
             nameNode(ParserKeys.Name).str,
             fullName,
             parserResult.filename,
             typeSpec.code,
             astParentType,
             astParentFullName
           )

         methodAstParentStack.push(typeDeclNode_)
         scope.pushNewScope(typeDeclNode_)
         val memberAsts = astForStructType(typeNode)

         methodAstParentStack.pop()
         scope.popScope()

         val modifier = addModifier(typeDeclNode_, nameNode(ParserKeys.Name).str)
         Seq(Ast(typeDeclNode_).withChild(Ast(modifier)).withChildren(memberAsts.toSeq))
    None => Seq.empty
}

hiteshbedre replied:
Addressed here: https://github.com/joernio/joern/pull/3567/commits/e7443b7c48e81e5924389917f0017b0a50503e55

pandurangpatil commented on joern-cli/frontends/gosrc2cpg/src/main/scala/io/joern/gosrc2cpg/astcreation/AstForExpressionCreator.scala:
```diff
@@ -23,24 +23,14 @@ trait AstForExpressionCreator(implicit withSchemaValidation: ValidationMode) { t
     }
   }
 
-  private def astForStructType(expr: ParserNodeInfo): Seq[Ast] = {
```
@hiteshbedre We can move this method as well to as this related to TypeDecl `AstForStructuresCreator.scala`
